### DISCUSSION TITLE
Re-rendering Server Components when props change

### DISCUSSION THREAD

**Discussion by lance-h** [vercel/next.js#50250] [App Router]
Let's say we have a ProductComponent which has a single prop, productId, which is the ID of a Product in a third party system. 

The ProductComponent performs a fetch request to a third-party API during rendering to retrieve the name of the product.

```
async function ProductComponent({ productId }: { productId: string }) {
  const product = await fetch(`https://productapi.com?product=${productId}`);
  const { name } = await product.json();

  return <div>{name}</div>;
}
```

This works as expected in that the API request is only made on the server-side.

**Is it possible to re-render / regenerate _just_ this one component when the props change on the client side?**

For example, Contentful has released a live preview / live update feature which uses react hooks (or other client-side-only logic) to merge pre-fetched data with updated data. In a situation like this where just a productId has changed it is not ideal/feasible to refresh the entire data used to render the current page. I realise this would involve a server trip (since we never want the product API call to happen on the client-side). We also only need this behaviour during "preview" mode (since we don't want to affect the performance on a live site).

**Comment by lance-h** [comment]
It seems like Next.js has the ability to inject server-side component HTML into the DOM during page navigation, however I can't see a way to  do this on demand.

**Comment by naseef0** [comment]
You can revalidate the entire page([where you want to update your cached data without waiting for a revalidation period to expire.](https://nextjs.org/docs/app/api-reference/functions/revalidatePath)) or you can revalidate where you want to update your cached data without waiting for a revalidation period to expire. ([https://nextjs.org/docs/app/api-reference/functions/revalidateTag](https://nextjs.org/docs/app/api-reference/functions/revalidateTag))

**Comment by lance-h** [comment]
Thanks @naseef0 that's helpful to know, however I'm not sure how I could use this knowledge to re-render a single server component without re-rendering the entire page.
Part of the problem I want to avoid is that re-rendering the entire page would mean that the function to retrieve the original props (from a CMS) would re-run. Due to caching of fetch calls the API call wouldn't be rerun, however I can't see a way to inject a value from the client side into a server-side component.

Perhaps I could rephrase the question slightly:
Is there a way to render a single server component on-demand with specific props?

I've expanded on the original example slightly to show the page-level server component which imports the "product" component.

I want to be able to re-render server ProductComponent (but no the ProductPageComponent) based on some client-side interaction.

```
// File: /app/product/[slug].tsx
//  Note: This file does not work since it's a server component with client-side logic, but I'm trying to use it for illustration

async function ProductPageComponent({ slug }: { slug: string }) {
  const productPage = await fetch(`https://mycms.com/productPage?slug=${slug}`);
  const { name, productId: originalProductId } = await productPage.json();
  
  const [productId, setProductId] = useState(originalProductId);

  return (<div>
    <h1>{name}</h1>
    <ProductComponent productId={productId} />
   <input type="text" onChange={(e) => { setProductId(e.target.value) }} />
    <button>Change Product</button>
  </div>);
}

// File: /components/product.tsx

async function ProductComponent({ productId }: { productId: string }) {
  // Note: I never want this call to occur on the client-side (hence why this is a server component)
  const product = await fetch(`https://productapi.com?product=${productId}`);
  const { price } = await product.json();

  return <div>${price}</div>;
}
```

**Reply by rajtomarejn** [reply]
+1 Facing same issue on `NextJS 14.0.3`. The **`server page always try to re-render and fetch new data just after the client interaction(serverAction) finished it's execution`**. My requirement and setup for the component is similar like above. If I'm not doing something wrong, then I think, this is not the right way to handle server data fetching as entire data and the component list re-renders on the user interaction. @lance-h will you be ok to help us with the resolution if you've got any.

**Comment by biwano** [comment]
+1

**Comment by dlopatiuk22** [comment]
@lance-h, @biwano , Please let me know if you have found a solution to this problem?

**Reply by TomJD** [reply]
Did you guys manage to get this to work? Literally trying to achieve the same, but I have no API routes rather I am fetching data using Prisma and Azure Data Tables directly from SQL DB and Table Storage

**Comment by naseef0** [comment]
@dlopatiuk22 @lance-h Did you guys tried https://nextjs.org/docs/app/api-reference/functions/server-actions?

**Reply by rajtomarejn** [reply]
Yes, I've tried. Not sure if I'm doing something extremely wrong.

**Comment by lance-h** [comment]
@dlopatiuk22 @rajtomarejn I don't have a resolution yet sorry. I have effectively avoided using the App router for now until it matures further as I was running into a number of critical issues. I'm still hopeful it will improve.

**Comment by sorenhoyer** [comment]
Good question. Hopefully Next.js will address this problem.
